### ISSUE TITLE
Improper metric computation in fit method (tf 2.0)

### ISSUE THREAD

**Issue by durandg12** [tensorflow/tensorflow#28338]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS 10.13.6
- TensorFlow installed from (source or binary): from pip I think, I don't remember and I don't know the difference
- TensorFlow version (use command below): v1.12.0-9492-g2c319fb415 2.0.0-alpha0
- Python version: 3.6.7

**Describe the current behavior**
When using the `fit` method on a custom model (the model in the guide [Writing layers and models with TensorFlow Keras](https://www.tensorflow.org/alpha/guide/keras/custom_layers_and_models#putting_it_all_together_an_end-to-end_example)), a simple `MeanSquaredError` metric seems to be improperly computed. When fitting, the metric is always greater than the loss, whereas the loss is computed as the mean squared error plus a custom error defined as a KL divergence in the model.

**Describe the expected behavior**
The mean squared error metric should always be lesser than the loss, as it is when training the model with custom code instead of using the `fit` method.

**Code to reproduce the issue**
 The following code is adapted from the guide [Writing layers and models with TensorFlow Keras](https://www.tensorflow.org/alpha/guide/keras/custom_layers_and_models#putting_it_all_together_an_end-to-end_example). I simply customized the formatting and added a metric to track the mean squared error in addition to the total loss.

On a side note, the code in [the guide](https://www.tensorflow.org/alpha/guide/keras/custom_layers_and_models#putting_it_all_together_an_end-to-end_example) seems to forget to use `reset_states` on the metric at the beginning of each epoch.

```
import tensorflow as tf
from tensorflow.keras import layers
from time import time

class Sampling(layers.Layer):
    # Uses (z_mean, z_log_var) to sample z, the vector encoding a digit.

    def call(self, inputs, training=None):
        z_mean, z_log_var = inputs
        batch = tf.shape(z_mean)[0]
        dim = tf.shape(z_mean)[1]
        epsilon = tf.keras.backend.random_normal(shape=(batch, dim))
        return z_mean + tf.exp(0.5 * z_log_var) * epsilon # = z_mean + tf.sqrt(z_var) * epsilon


class Encoder(layers.Layer):
    # Maps MNIST digits to a triplet (z_mean, z_log_var, z).

    def __init__(self,
                 latent_dim=32,
                 intermediate_dim=64,
                 name='encoder',
                 **kwargs):
        super(Encoder, self).__init__(name=name, **kwargs)
        self.dense_proj = layers.Dense(intermediate_dim, activation='relu')
        self.dense_mean = layers.Dense(latent_dim)
        self.dense_log_var = layers.Dense(latent_dim)
        self.sampling = Sampling()

    def call(self, inputs, training=None):
        x = self.dense_proj(inputs)
        z_mean = self.dense_mean(x)
        z_log_var = self.dense_log_var(x)
        z = self.sampling((z_mean, z_log_var))
        return z_mean, z_log_var, z


class Decoder(layers.Layer):
    # Converts z, the encoded digit vector, back into a readable digit.

    def __init__(self,
                 original_dim,
                 intermediate_dim=64,
                 name='decoder',
                 **kwargs):
        super(Decoder, self).__init__(name=name, **kwargs)
        self.dense_proj = layers.Dense(intermediate_dim, activation='relu')
        self.dense_output = layers.Dense(original_dim, activation='sigmoid')

    def call(self, inputs, training=None):
        x = self.dense_proj(inputs)
        return self.dense_output(x)


class VariationalAutoEncoder(tf.keras.Model):
    # Combines the encoder and decoder into an end-to-end model for training.

    def __init__(self,
                 original_dim,
                 intermediate_dim=64,
                 latent_dim=32,
                 name='autoencoder',
                 **kwargs):
        super(VariationalAutoEncoder, self).__init__(name=name, **kwargs)
        self.original_dim = original_dim
        self.encoder = Encoder(latent_dim=latent_dim,
                               intermediate_dim=intermediate_dim)
        self.decoder = Decoder(original_dim=original_dim,
                               intermediate_dim=intermediate_dim)

    def call(self, inputs, training=None):
        z_mean, z_log_var, z = self.encoder(inputs)
        reconstructed = self.decoder(z)
        # Add KL divergence regularization loss.
        kl_loss = - 0.5 * tf.reduce_mean(
            z_log_var - tf.square(z_mean) - tf.exp(z_log_var) + 1)
        self.add_loss(kl_loss)
        return reconstructed


# Data preparation
original_dim = 28 * 28
(x_train, _), _ = tf.keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, original_dim).astype('float32') / 255
train_dataset = tf.data.Dataset.from_tensor_slices(x_train)
train_dataset = train_dataset.shuffle(buffer_size=1024).batch(64)

# Model creation
vae = VariationalAutoEncoder(original_dim, intermediate_dim=64, latent_dim=32)
optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)
mse_loss_fn = tf.keras.losses.MeanSquaredError()
loss_metric = tf.keras.metrics.Mean()
mse_metric = tf.keras.metrics.Mean()

@tf.function
def train_step(inputs):
    with tf.GradientTape() as tape:
          reconstructed = vae(inputs)
          # Compute reconstruction loss
          mse_loss = mse_loss_fn(inputs, reconstructed)
          loss = mse_loss + tf.add_n(vae.losses)  # Add KLD regularization loss
    
    grads = tape.gradient(loss, vae.trainable_variables)
    optimizer.apply_gradients(zip(grads, vae.trainable_variables))
    # Update the metrics
    loss_metric.update_state(loss)
    mse_metric.update_state(mse_loss)

def train_over_epochs(train_dataset, epochs):
    # Iterate over epochs.
    for epoch in range(epochs):
        begin_time = time()
        print('Start of epoch {:d}'.format(epoch + 1))
        # Reset the metrics
        loss_metric.reset_states()
        mse_metric.reset_states()
    
        # Iterate over the batches of the dataset.
        for step, x_batch_train in enumerate(train_dataset):
            train_step(x_batch_train)
        
        print('    step {:3d}: mean total loss = {:f}'.format(step + 1, loss_metric.result().numpy()))
        print('              mean mse loss = {:f}'.format(mse_metric.result().numpy()))
        print('    time: {:.2f}'.format(time() - begin_time))
        
train_over_epochs(train_dataset, epochs=4)
```
The output of this code is something like this:
```
Start of epoch 1
    step 938: mean total loss = 0.074619
              mean mse loss = 0.074219
    time: 4.24
Start of epoch 2
    step 938: mean total loss = 0.067608
              mean mse loss = 0.067607
    time: 2.51
Start of epoch 3
    step 938: mean total loss = 0.067527
              mean mse loss = 0.067526
    time: 2.64
Start of epoch 4
    step 938: mean total loss = 0.067471
              mean mse loss = 0.067470
    time: 2.50

```
The mse loss is always smaller than the total loss which is expected.

Now if we run the following code:
```
train_for_fit_dataset=train_dataset.map(lambda x: (x,x)) #create labels equal to the data for fit method
vae = VariationalAutoEncoder(original_dim, intermediate_dim=64, latent_dim=32)
vae.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3), 
           loss=tf.keras.losses.MeanSquaredError(),
           metrics=[tf.keras.metrics.MeanSquaredError()])
vae.fit(train_for_fit_dataset, epochs=4)
```
The output looks like this:
```
Epoch 1/4
938/938 [==============================] - 5s 5ms/step - loss: 0.0744 - mean_squared_error: 0.0923
Epoch 2/4
938/938 [==============================] - 4s 4ms/step - loss: 0.0676 - mean_squared_error: 0.0678
Epoch 3/4
938/938 [==============================] - 4s 4ms/step - loss: 0.0675 - mean_squared_error: 0.0677
Epoch 4/4
938/938 [==============================] - 4s 5ms/step - loss: 0.0675 - mean_squared_error: 0.0676
```
Note that the loss seems be computed as in the previous code, but the mean squared error is definitely not.

Either there is a subtlety in `fit` I am not aware of, either there is a bug.

**Comment by durandg12** [comment]
I have updated the code I provided because in my custom training loop I was updating `mse_metric` after applying gradients whereas I was updating `loss_metric` with the loss value computed before applying them. Now I consistently update both at the same time.

I wonder if such a timing consideration is likely to explain the results observed with the `fit` method. The larger gap at epoch 1 is still mysterious.
